File: Sort/4MedianOfTwoSortedArrays.py
# Binary Search Solution
# 他的优点是够快 O(log(min(n, m)))
# 把两个数组分成两部分, 然后比较两个数组的中间值, 
# 如果两个数组的中间值相等, 那么就是中位数, 
# 如果不相等, 那么就把中间值小的那个数组的左半部分和中间值大的那个数组的右半部分合并, 然后再找中间值, 直到找到中位数
# 今天太累了 下次看吧 ):
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        n1 = len(nums1)
        n2 = len(nums2)

        # Ensure nums1 is the smaller array for simplicity
        if n1 > n2:
            return self.findMedianSortedArrays(nums2, nums1)

        n = n1 + n2
        left = (n1 + n2 + 1) // 2 # Calculate the left partition size
        low = 0
        high = n1

        while low <= high:
            mid1 = (low + high) // 2 # Calculate mid index for nums1
            mid2 = left - mid1 # Calculate mid index for nums2

            l1 = float('-inf')
            l2 = float('-inf')
            r1 = float('inf')
            r2 = float('inf')

            # Determine values of l1, l2, r1, and r2
            if mid1 < n1:
                r1 = nums1[mid1]
            if mid2 < n2:
                r2 = nums2[mid2]
            if mid1 - 1 >= 0:
                l1 = nums1[mid1 - 1]
            if mid2 - 1 >= 0:
                l2 = nums2[mid2 - 1]

            if l1 <= r2 and l2 <= r1:
                # The partition is correct, we found the median
                if n % 2 == 1:
                    return max(l1, l2)
                else:
                    return (max(l1, l2) + min(r1, r2)) / 2.0
            elif l1 > r2:
                # Move towards the left side of nums1
                high = mid1 - 1
            else:
                # Move towards the right side of nums1
                low = mid1 + 1

        return 0 # If the code reaches here, the input arrays were not sorted.

File: Array/4MedianOfTwoSortedArrays.java
// # Binary Search Solution
// # 他的优点是够快 O(log(min(n, m)))
// # 把两个数组分成两部分, 然后比较两个数组的中间值, 
// # 如果两个数组的中间值相等, 那么就是中位数, 
// # 如果不相等, 那么就把中间值小的那个数组的左半部分和中间值大的那个数组的右半部分合并, 然后再找中间值, 直到找到中位数
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;

        if (m > n) { // ensure num1 is smaller
            return findMedianSortedArrays(nums2, nums1);
        }

        int left = 0;
        int right = m;

        while (left <= right) {
            int partitionA = (left + right) / 2; // mid for left
            int partitionB = (m + n + 1) / 2 - partitionA; // mid for right

            int maxLeftA = (partitionA == 0)
                    ? Integer.MIN_VALUE
                    : nums1[partitionA - 1];
            int minRightA = (partitionA == m)
                    ? Integer.MAX_VALUE
                    : nums1[partitionA];
            int maxLeftB = (partitionB == 0)
                    ? Integer.MIN_VALUE
                    : nums2[partitionB - 1];
            int minRightB = (partitionB == n)
                    ? Integer.MAX_VALUE
                    : nums2[partitionB];

            if (maxLeftA <= minRightB && maxLeftB <= minRightA) {
                if ((m + n) % 2 == 0) {
                    return ((Math.max(maxLeftA, maxLeftB) +
                            Math.min(minRightA, minRightB)) /
                            2.0);
                } else {
                    return Math.max(maxLeftA, maxLeftB);
                }
            } else if (maxLeftA > minRightB) {
                right = partitionA - 1;
            } else {
                left = partitionA + 1;
            }
        }
        return 0.0;
    }
}